testdriver/testdriver.go

```diff
@@ -538,46 +525,148 @@ func (d *TestDriver) listenDriver() error {
+	if err != nil {
+		ef("%v", err)
 	}
-	log := func(format string, args ...interface{}) {
-		fmt.Fprintf(l, "[vim test client] "+format, args...)
+	if resp[0] != "" {
+		// this is a protocol-level error
+		ef("got error response: %v", resp[0])
 	}
-	log("logging enabled")
-
-	defer func() {
-		r := recover()
-		if r == nil {
-			return
+	// resp[1] will be a []intferface{} where the first
+	// element will be a Vim-level error
+	vimResp := resp[1].([]interface{})
+	if err := vimResp[0].(string); err != "" {
+		// this was a vim-level error
+		if !*bang {
+			ef("unexpected command error: %v", err)
+		}
+		fmt.Fprintln(os.Stderr, err)
+	}
+	if len(vimResp) == 2 {
+		if *bang {
+			ef("unexpected command success")
+		}
+		if *stringout {
+			switch vimResp[1].(type) {
+			case string:
+				fmt.Print(vimResp[1])
+			default:
+				ef("response type is %T, not string", vimResp[1])
+			}
+		} else {
+			enc := json.NewEncoder(os.Stdout)
+			if *indent {
+				enc.SetIndent("", "  ")
+			}
+			if err := enc.Encode(vimResp[1]); err != nil {
+				ef("failed to format output of JSON: %v", err)
+			}
 		}
-		exitCode = -1
-		fmt.Fprintln(os.Stderr, r)
-		log("panic with error: %+v", r)
-	}()
-
-	ef := func(format string, args ...interface{}) {
-		log(format, args...)
-		panic(fmt.Sprintf(format, args...))
 	}
+	return 0
+}
+
+func VimExprWait() (exitCode int) {
+	defer cleanUp(&exitCode)
 
 	fs := flag.NewFlagSet("vim", flag.PanicOnError)
-	bang := fs.Bool("bang", false, "expect command to fail")
-	indent := fs.Bool("indent", false, "pretty indent resulting JSON")
+	noindent := fs.Bool("noindent", false, "do not pretty indent resulting JSON")
 	stringout := fs.Bool("stringout", false, "print resulting string rather than JSON encoded version of string")
-
-	log("starting vim driver client and parsing flags...")
+	fWait := fs.String("wait", "", "retry (with exp backoff) until this time period has elapsed")
 
 	fs.Parse(os.Args[1:])
 	args := fs.Args()
+	if len(args) < 2 {
+		ef("invalid arguments")
+	}
+	fileName := args[0]
+	want, err := ioutil.ReadFile(fileName)
+	if err != nil {
+		ef("failed to read %v: %v", fileName, err)
+	}
+	// Now repeatedly evaluate the provided expression
+	// with exp backoff until we match want. Abort on any errors
+	// because this might indicative of a greater problem.
+	vimCmd := append([]string{"expr"}, args[1:]...)
+	if *fWait == "" {
+		fWait = &DefaultErrLogMatchWait
+	}
+	wait, err := time.ParseDuration(*fWait)
+	if err != nil {
+		ef("failed to parse -wait duration %q: %v", *fWait, err)
+	}
+	strategy := retry.LimitTime(wait,
+		retry.Exponential{
+			Initial: 100 * time.Millisecond, // be slightly less aggressive with Vim
+			Factor:  1.5,
+		},
+	)
+	var got *bytes.Buffer
+	for a := retry.Start(strategy, nil); a.Next(); {
+		resp, err := vim(vimCmd)
+		if err != nil {
+			ef("%v", err)
+		}
+		if resp[0] != "" {
+			ef("protocol error: %v", resp[0])
+		}
+		vimResp := resp[1].([]interface{})
```

Comment:
Not sure if this is possible, but could you push down this type coercion into the vim func, and return a better type?